Domex wavefront

CFGrote opened the pull request:
This PR implements a [domain extension for coherent wavefront propagation](https://github.com/PaNOSC-ViNYL/ViNYL-project/issues/16)
*Implements issue:* https://github.com/PaNOSC-ViNYL/ViNYL-project/issues/16

# Will add this information later.

## Description

*What is introduced, removed or renamed and why?*
*What is made required, recommended, optional?*
*What concept stands behind this change?*

*Please also add an example.*

## Affected Components

- `base`
- `EXT-...`

## Logic Changes

*Which logic changes are conceptually introduced?*

## Writer Changes

*How does this change affect data writers?*
*What would a writer need to change?*
*Does this pull request change the interpretation of existing data writers?*

- `libopenPMD` (upcoming): https://github.com/ComputationalRadiationPhysics/libopenPMD/...

## Reader Changes

*How does this change affect data readers?*

*What would a reader need to change? Link implementation examples!*

- `openPMD-validator`: https://github.com/openPMD/openPMD-validator/...
- `openPMD-viewer`: https://github.com/openPMD/openPMD-viewer/...
- `yt`: https://github.com/yt-project/yt/...
- `VisIt`: https://github.com/openPMD/openPMD-visit-plugin
- `libopenPMD` (upcoming): https://github.com/ComputationalRadiationPhysics/libopenPMD/...
- `XDMF` (upcoming): https://github.com/openPMD/openPMD-tools/...

## Data Converter

*How does this affect already existing files with previous versions of the standard?*
*Is this change possible to be expressed in a way, that an automated tool can update the file?*
*Link code/pull requests that implement the upgrade logic!*

- https://github.com/openPMD/openPMD-converter/...

CFGrote committed:
First draft for a domain extension for wavefront data.

CFGrote committed:
Adding more required attributes and correct unitDimensions.

CFGrote committed:
Adding more required attributes and correct unitDimensions.

CFGrote committed:
Add cartesian components

CFGrote committed:
Adding units to photon energy

CFGrote committed:
Update EXT_WAVEFRONT.md

CFGrote committed:
Update EXT_WAVEFRONT.md

Series attributes

CFGrote committed:
Update standard for waves

Restructuring attributes and move physical data to series

JunCEEE commented on EXT_WAVEFRONT.md:
```diff
@@ -0,0 +1,122 @@
+    - scope: *required*
+
+- `photon energy`
+    - type: *(floatX)* or *(intX)* or *(uintX)*
```
On which occasion the energy has `int` type?

CFGrote replied:
this is just for convenience, so photon_energy=4000 (4 keV) would not raise an error.

JunCEEE commented on EXT_WAVEFRONT.md:
```diff
@@ -0,0 +1,122 @@
+
+- fundamental fields:
+  - `E_real/x` and `E_real/y`
+    - type: *(floatX)* or *(intX)* or *(uintX)*
```
The same question as that of the energy. The `int` type confuses me. On which occasion the field has `int` type?

JunCEEE commented on EXT_WAVEFRONT.md:
```diff
@@ -0,0 +1,122 @@
+                              if attribute `Fourier domain` is 'time', or
+                              `unitDimension = (0., -1,0, 0., 0., 0., 0., 0.)`
+                              ((J / eV)^{1/2} / m  = m^{-1})
+                              if attribute `Fourier domain` is 'frequency`.
```
`Fourier domain` or `temporal domain`?

CFGrote replied:
temporal domain is more accurate, thanks for catching

JunCEEE commented on EXT_WAVEFRONT.md:
```diff
@@ -0,0 +1,122 @@
+                              if attribute `Fourier domain` is 'time', or
+                              `unitDimension = (0., -1,0, 0., 0., 0., 0., 0.)`
+                              ((J / eV)^{1/2} / m  = m^{-1})
+                              if attribute `Fourier domain` is 'frequency`.
```
`Fourier domain` or `temporal domain`?

JunCEEE reviewed (commented):
One obvious problem: `Fourier domain` is indeed undefined in this standard, but it appears in the `mesh records` definition, I assume it might be replaced by `temporal domain`.

I am not an expert in this field. So there are not many comments, but just some naive questions among the file.

CFGrote committed:
Update EXT_WAVEFRONT.md

resolves @ejcjason comment on PR #5 (https://github.com/PaNOSC-ViNYL/openPMD-standard/pull/5/files/efd45d622d7eafa1b5133080965ed3975d67d9ba#r339889356)

CFGrote committed:
Update EXT_WAVEFRONT.md

Resolves @ejcjason comment on #5 here: https://github.com/PaNOSC-ViNYL/openPMD-standard/pull/5/files/efd45d622d7eafa1b5133080965ed3975d67d9ba#r339889356

JunCEEE commented:
At line 112 and 117, there are still `Fourier domain`, should they also be changed to `temporal domain` for consistency?
